lib/cretonne/src/flowgraph.rs

```diff
@@ -167,10 +168,108 @@ impl ControlFlowGraph {
         postorder
     }
 
+    /// Return ebbs from a block in post-order, starting from an entry point in the block.
+    pub fn postorder_ebbs_block(&self, entry_block: &Ebb, blocks: &HashSet<Ebb>) -> Vec<Ebb> {
+        let mut grey = HashSet::new();
+        let mut black = HashSet::new();
+        let mut stack = vec![entry_block.clone()];
+        let mut postorder = Vec::new();
+
+        while !stack.is_empty() {
+            let node = stack.pop().unwrap();
+            if !grey.contains(&node) {
+                // This is a white node. Mark it as gray.
+                grey.insert(node);
+                stack.push(node);
+                // Get any children we've never seen before.
+                for child in self.get_successors(node) {
+                    if blocks.contains(child) && !grey.contains(child) {
+                        stack.push(child.clone());
+                    }
+                }
+            } else if !black.contains(&node) {
+                postorder.push(node.clone());
+                black.insert(node.clone());
+            }
+        }
+        postorder
+    }
+
     /// An iterator across all of the ebbs stored in the CFG.
     pub fn ebbs(&self) -> Keys<Ebb> {
         self.data.keys()
     }
+
+    /// Implementation of Tarjan's strongly connected components algorithm.
+    pub fn tarjan_scc(&self) -> Vec<HashSet<Ebb>> {
+        let index: u32 = 0;
+        let mut stack: Vec<Ebb> = Vec::new();
+        let mut components: Vec<HashSet<Ebb>> = Vec::new();
+        // properties are (index,lowlink,onStack) for each vertex
+        let mut properties: HashMap<Ebb, (u32, u32, bool)> = HashMap::new();
+        match self.entry_block {
+            None => (),
+            Some(node) => {
+                self.strong_connect(&node, &mut stack, &mut components, &mut properties, index)
+            }
+        }
+        components
+    }
+
+    // Helper recursive function for the Tarjan's SCC algorithm
+    fn strong_connect(&self,
+                      node: &Ebb,
+                      stack: &mut Vec<Ebb>,
+                      components: &mut Vec<HashSet<Ebb>>,
+                      properties: &mut HashMap<Ebb, (u32, u32, bool)>,
+                      index: u32) {
+        properties.insert(node.clone(), (index, index, true));
+        stack.push(node.clone());
```

Comment:
FWIW, [Nuutila et al](http://citeseerx.ist.psu.edu/viewdoc/summary?doi=10.1.1.53.5911) give some minor improvements to Tarjan's original algorithm. It's not urgent that we implement these now, but it would be nice to mention them in a TODO comment or so until we get to them.